swarm/src/protocols_handler/multi.rs

```diff
@@ -0,0 +1,362 @@
+    }
+
+    fn poll(&mut self, cx: &mut Context)
+        -> Poll<ProtocolsHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::OutEvent, Self::Error>>
+    {
+        // Not always polling handlers in the same order should give anyone the chance to make progress.
+        let pos = rand::thread_rng().gen_range(0, self.handlers.len());
+
+        for (k, h) in self.handlers.iter_mut().skip(pos) {
+            if let Poll::Ready(e) = h.poll(cx) {
+                let e = e.map_outbound_open_info(|i| (k.clone(), i)).map_custom(|p| (k.clone(), p));
+                return Poll::Ready(e)
+            }
+        }
+
+        for (k, h) in self.handlers.iter_mut().take(pos) {
+            if let Poll::Ready(e) = h.poll(cx) {
+                let e = e.map_outbound_open_info(|i| (k.clone(), i)).map_custom(|p| (k.clone(), p));
+                return Poll::Ready(e)
+            }
+        }
+
+        Poll::Pending
+    }
+}
+
+/// A [`IntoProtocolsHandler`] for multiple other `IntoProtocolsHandler`s.
+#[derive(Clone)]
+pub struct IntoMultiHandler<K, H> {
+    handlers: HashMap<K, H>
+}
+
+impl<K, H> fmt::Debug for IntoMultiHandler<K, H>
+where
+    K: fmt::Debug + Eq + Hash,
+    H: fmt::Debug
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("IntoMultiHandler")
+            .field("handlers", &self.handlers)
+            .finish()
+    }
+}
+
+
+impl<K, H> IntoMultiHandler<K, H>
+where
+    K: Hash + Eq,
+    H: IntoProtocolsHandler
+{
+    pub fn try_from_iter<I>(iter: I) -> Result<Self, DuplicateProtoname>
+    where
+        I: IntoIterator<Item = (K, H)>
+    {
+        let m = IntoMultiHandler { handlers: HashMap::from_iter(iter) };
+        uniq_proto_names(m.handlers.values().map(|h| h.inbound_protocol()))?;
+        Ok(m)
+    }
+}
+
+impl<K, H> IntoProtocolsHandler for IntoMultiHandler<K, H>
+where
+    K: Clone + fmt::Debug + Eq + Hash + Send + 'static,
+    H: IntoProtocolsHandler
+{
+    type Handler = MultiHandler<K, H::Handler>;
+
+    fn into_handler(self, p: &PeerId, c: &ConnectedPoint) -> Self::Handler {
+        MultiHandler {
+            handlers: self.handlers.into_iter()
+                .map(|(k, h)| (k, h.into_handler(p, c)))
+                .collect()
+        }
+    }
+
+    fn inbound_protocol(&self) -> <Self::Handler as ProtocolsHandler>::InboundProtocol {
+        Upgrade {
+            upgrades: self.handlers.iter()
+                .map(|(k, h)| (k.clone(), h.inbound_protocol()))
+                .collect()
+        }
+    }
+}
+
+/// Key and protocol name pair used as `UpgradeInfo::Info`.
+#[derive(Debug, Clone)]
+pub struct KeyedProtoName<K, H>(K, H);
+
+impl<K, H: ProtocolName> ProtocolName for KeyedProtoName<K, H> {
+    fn protocol_name(&self) -> &[u8] {
+        self.1.protocol_name()
+    }
+}
+
+/// Inbound and outbound upgrade for all `ProtocolsHandler`s.
+#[derive(Clone)]
+pub struct Upgrade<K, H> {
+    upgrades: HashMap<K, H>
```

Comment:
I feel like this `HashMap` here is overkill.

```suggestion
    upgrades: Vec<(K, H)>
```

Can we not index the upgrades with a simple `usize` rather than with `K`?

Reply:
What would be the advantage of restricting ourselves to `usize`? The substrate PR for instance, where this originates from, uses the `ProtocolId` for indexing.

Reply:
Note that I'm talking only about `Upgrades` here. The `MultiHandler` should keep a `HashMap`.

If we compare:
With a `HashMap` in `Upgrades`:

- When a substream is opened by the remote, we go through each child handler, extract their inbound upgrade, clone the keys, and create a `HashMap`.
- The list of protocols is then gathered, and we clone the keys again.
- When a protocol is negotiated, we index the hashmap for the corresponding key.

With a `Vec` in `Upgrades`:

- When a substream is opened by the remote, we go through each child handler, extract their inbound upgrade, clone the keys, and create a `Vec`.
- The list of protocols is then gathered, and we simply put `usize`s in the upgrades. We don't clone the key.
- When a protocol is negotiated, we look in the `Vec` for corresponding index.

Overall the `Vec` solution looks overall better as we don't clone the keys, and we don't go through the overhead of a `HashMap`.

Reply:
Clients which are worried about expensive `Clone`s can use a `usize` or some other "cheap" type directly.

Using `usize` only in `Upgrades` requires another `K` <-> `usize` mapping because the output of an upgrade is given to handler methods where we need to map back from `usize` to `K` in order to get back the corresponding handler from the `HashMap` of `MultiHandler`.

I am fine with using a `Vec<(K, H)>` even though I would rather like to see this presumed overhead to be quantified.

Reply:
Ah I see from your changes that we didn't think of the same thing.

To me the idea is that `protocol_info` should become:
```rust
    fn protocol_info(&self) -> Self::InfoIter {
        self.upgrades.iter().enumerate().map(|(n, (_, i))| iter::repeat(n).zip(i.protocol_info()))
            .flatten()
            .map(|(n, i)| KeyedProtoName(n, i))
            .collect::<Vec<_>>()
            .into_iter()
    }
```

And that `KeyedProtoName` should become:
```rust
pub struct KeyedProtoName<H>(usize, H);
```